umbra-js/src/classes/Umbra.ts

```diff
@@ -235,6 +256,119 @@ export class Umbra {
     return { tx, stealthKeyPair };
   }
 
+  async batchSend(signer: JsonRpcSigner | Wallet, batch: SendBatch[], overrides: SendOverrides = {}) {
+    let tokenSum = new Map();
+    let tokens: string[] = [];
+    for (let i = 0; i < batch.length; i++) {
```

Comment:
This loop and the next loop are trying to do a lot of things: validate recipient address, compute total sums of tokens, get an array of unique tokens, and check balances. We can make things a bit cleaner here, and _a lot_ more performant by taking the `await`s out of the loops

The code below renames the `batch` input to something plural like `sends`. The code should be considered pseduocode, may not work but conveys the idea. Lmk what you think of something like this:

```typescript
async batchSend(signer: JsonRpcSigner | Wallet, sends: SendBatch[], overrides: SendOverrides = {}) {
  // Get all unique recipients and validate them
  const recipients = new Set(sends.map(send => send.recipient))
  await Promise.all(recipients.map(assertSupportedAddress))

  // Get a unique list of tokens
  const tokens = new Set(sends.map(send => send.token))

  // Validate balances
  //    - pull out the validation logic re-used by both these functions into a helper method, say
  //      it's called `assertSufficientBalance`
  //    - this line assumes you created a `balances` mapping somewhere above, perhaps
  //      using `for (const send of sends) {` 
  await Promise.all(tokens.map(token => assertSufficientBalance(balances.get(token)))

  // --- snip ---
}
```

Reply:
Definitely cleaner, but my implementation still uses for loops. Using `Promise.all` seem to just add an extra step but maybe I'm wrong? 082acd9876681cd3c0488d5f4d824b8af1507f88

Reply:
https://github.com/ScopeLift/umbra-protocol/blob/1b968b7f98e2a8dc3a68e91dbe806bde65b17568/umbra-js/src/classes/Umbra.ts#L218-L221

Like you said, the above still uses a for loop. This isn't great because, say you have 10 addresses. Each iteration of the for loop `await`s a response, and only after 10 serial awaits can we move on. This is slow and therefore bad UX (`await` in a for loop is typically a code smell that something is suboptimal and should be refactored)

If you replace the for loop with:

```typescript
const assertionPromises = recipients.map(assertSupportedAddress);
await Promise.all(assertionPromises);

// Or, identical, just merged into one line
await Promise.all(recipients.map(assertSupportedAddress))

// Also identical, maybe more clear if not super familiar with callbacks
const assertionPromises = recipients.map(recipient => assertSupportedAddress(recipient));
await Promise.all(assertionPromises);
```

then the `map` call dispatches the async calls for all promises at (roughly) the same time, instead of sending each one and waiting for a response one by one in a loop. Then we just wait for all of them to resolve with `Promise.all`. Because all async calls happen in parallel, and we wait for them all in parallel, it's much faster. 

--------

Same concept for the below block. Let's remove all `await`s in a loop here. The main reason for having `await`s in a loop is when e.g. iteration 2 needs the result of iteration 1, but that's not the case here

https://github.com/ScopeLift/umbra-protocol/blob/1b968b7f98e2a8dc3a68e91dbe806bde65b17568/umbra-js/src/classes/Umbra.ts#L231-L234

Reply:
Awesome! Thanks so much for the explanation, I'm learning so much here. Does this look good? 66133c10cc602fd3cced862974f3b89b05211f18

Reply:
Yes! Looks really good. Just one more `await` in a loop when you call `this.prepareSend`

Reply:
How's this looking? https://github.com/ScopeLift/umbra-protocol/pull/426/commits/fe2a711959650aa19f34b01312093f022dbcf740